### PR TITLE
IntersectionObserver

### PR DESCRIPTION
**Pull request by triblondon**
Adds support for IntersectionObserver using polyfill developed by @philipwalton and @surma.

### CODE REVIEW

File: polyfills/IntersectionObserver/polyfill.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
+/*
```
Comment (JakeChampion) [inline polyfills/IntersectionObserver/polyfill.js]: If this is published on NPM then we could use our update tasks to pull in the polyfill

Comment (triblondon) [inline polyfills/IntersectionObserver/polyfill.js]: Yup, I suggested to @surma that he publish on npm.

File: polyfills/IntersectionObserver/polyfill.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    this._rootMargin = options.rootMargin || [0, 0, 0, 0];
+    this._thresholds = options.threshold || [0];
+    this._init();
+  };
+
+  IntersectionObserver.prototype = {
+    //
+    // Public API
+    //
+    get root() {
```
Comment (JakeChampion) [inline polyfills/IntersectionObserver/polyfill.js]: I believe this is ES6/2015 method definition syntax. Do we need to change this to ES5?

Comment (surma) [inline polyfills/IntersectionObserver/polyfill.js]: That’s ES5 :)

File: polyfills/IntersectionObserver/polyfill.js (hunk 1 above)
Comment (surma) [inline polyfills/IntersectionObserver/polyfill.js]: I have no strong opinions on where or how to publish this. NPM seems like a good idea, I was just told to PR this into the WICG/IO repo :)

**Comment by triblondon** [comment]
Well, the good news is the Mocha upgrade didn't break all the tests.  The bad news is IO is still failing on CI in every browser...

```
- firefox 30 (Sauce results: https://saucelabs.com/tests/7b9fd71ab31944259db97a7bb81de360)
    -> IntersectionObserver (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=IntersectionObserver (timeout after 10000ms, 2 retries)
 - ie 10 (Sauce results: https://saucelabs.com/tests/2d7d7854a017429cae65763788e3dbfc)
    -> IntersectionObserver (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=IntersectionObserver (timeout after 10000ms, 2 retries)
 - ie 8 (Sauce results: https://saucelabs.com/tests/30442fd52afc4555a57949cd23a8c1c9)
    -> IntersectionObserver (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=IntersectionObserver (timeout after 10000ms, 2 retries)
 - ie 7 (Sauce results: https://saucelabs.com/tests/174b4605f43245f38588be4980254500)
    -> IntersectionObserver (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=IntersectionObserver (timeout after 10000ms, 2 retries)
 - safari 9 (Sauce results: https://saucelabs.com/tests/8d6bde2276974d91882748675d701fcf)
    -> IntersectionObserver (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=IntersectionObserver (timeout after 10000ms, 2 retries)
 - safari 8 (Sauce results: https://saucelabs.com/tests/ad5db24cc1004f6f992afca6c4ca1ccd)
    -> IntersectionObserver (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=IntersectionObserver (timeout after 10000ms, 2 retries)
 - android 4.4 (Sauce results: https://saucelabs.com/tests/9a16f4b275d2422fb8f86a543b43ac74)
    -> IntersectionObserver (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=IntersectionObserver (timeout after 10000ms, 2 retries)
```

**Comment by triblondon** [comment]
OK, looks like Sauce can't load Sinon?
https://saucelabs.com/tests/c505172bd62a4f2798ead79483a97042

image

**Comment by triblondon** [comment]
Fixed that.  Now there's a simple expectation failure...

image

**Comment by triblondon** [comment]
Posted https://github.com/WICG/IntersectionObserver/pull/135#issuecomment-226473998 to query test failures with Google team.

**Comment by triblondon** [comment]
It works!

**Comment by triblondon** [comment]
@JakeChampion are you happy for us to line this up for 3.8?

**Comment by JakeChampion** [comment]
Yes, definitely!

**Comment by JakeChampion** [comment]
🎉

**Pull request merged**